addon/-private/system/model/internal-model.js

```diff
@@ -278,8 +290,22 @@ export default class InternalModel {
     this._triggerDeferredTriggers();
   }
 
-  recordObjectWillDestroy() {
+  resetRecord() {
     this.record = null;
+    this.dataHasInitialized = false;
+    this.isReloading = false;
+    this.error = null;
+    this.currentState = RootState.empty;
+    this.__attributes = null;
+    this.__inFlightAttributes = null;
+  }
+
+  dematerializeRecord() {
+    if (this.record) {
```

Comment:
The case where we wind up here with `!this.record` is `store.unloadAll(type)` which is public API.

We could change this to assert `!!this.record` and then tweak the store to only unload materialized internal models.  thoughts?